contracts/services/reverse-splitter/src/msg.rs

```diff
@@ -0,0 +1,424 @@
+}
+
+#[cw_serde]
+pub enum UncheckedSplitAmount {
+    FixedAmount(Uint128),
+    FixedRatio(Decimal),
+    DynamicRatio {
+        contract_addr: String,
+        params: String,
+    },
+}
+
+#[cw_serde]
+pub struct UncheckedSplitConfig {
+    pub denom: UncheckedDenom,
+    pub account: ServiceAccountType,
+    pub amount: UncheckedSplitAmount,
+    pub factor: Option<u64>,
+}
+
+impl UncheckedSplitConfig {
+    pub fn new(
+        denom: UncheckedDenom,
+        account: impl Into<ServiceAccountType>,
+        amount: UncheckedSplitAmount,
+        factor: Option<u64>,
+    ) -> Self {
+        UncheckedSplitConfig {
+            denom,
+            account: account.into(),
+            amount,
+            factor,
+        }
+    }
+
+    pub fn with_native_amount(amount: u128, denom: &str, input: &Addr) -> Self {
+        UncheckedSplitConfig::new(
+            UncheckedDenom::Native(denom.to_string()),
+            input,
+            UncheckedSplitAmount::FixedAmount(amount.into()),
+            None,
+        )
+    }
+
+    pub fn with_cw20_amount(amount: u128, addr: &Addr, input: &Addr) -> Self {
+        UncheckedSplitConfig::new(
+            UncheckedDenom::Cw20(addr.to_string()),
+            input,
+            UncheckedSplitAmount::FixedAmount(amount.into()),
+            None,
+        )
+    }
+
+    pub fn with_native_ratio(ratio: Decimal, denom: &str, input: &Addr) -> Self {
+        UncheckedSplitConfig::new(
+            UncheckedDenom::Native(denom.to_string()),
+            input,
+            UncheckedSplitAmount::FixedRatio(ratio),
+            None,
+        )
+    }
+
+    pub fn with_cw20_ratio(ratio: Decimal, addr: &Addr, input: &Addr) -> Self {
+        UncheckedSplitConfig::new(
+            UncheckedDenom::Cw20(addr.to_string()),
+            input,
+            UncheckedSplitAmount::FixedRatio(ratio),
+            None,
+        )
+    }
+
+    pub fn with_native_dyn_ratio(
+        contract_addr: &Addr,
+        params: &str,
+        denom: &str,
+        input: &Addr,
+    ) -> Self {
+        UncheckedSplitConfig::new(
+            UncheckedDenom::Native(denom.to_string()),
+            input,
+            UncheckedSplitAmount::DynamicRatio {
+                contract_addr: contract_addr.to_string(),
+                params: params.to_string(),
+            },
+            None,
+        )
+    }
+
+    pub fn with_cw20_dyn_ratio(
+        contract_addr: &Addr,
+        params: &str,
+        addr: &Addr,
+        input: &Addr,
+    ) -> Self {
+        UncheckedSplitConfig::new(
+            UncheckedDenom::Cw20(addr.to_string()),
+            input,
+            UncheckedSplitAmount::DynamicRatio {
+                contract_addr: contract_addr.to_string(),
+                params: params.to_string(),
+            },
+            None,
+        )
+    }
+
+    pub fn with_factor(mut self, factor: u64) -> Self {
+        self.factor = Some(factor);
+        self
+    }
```

Comment:
The rest of the helper "with_" function are creating a new struct, I would suggest calling this as "set_factor" instead of with to make it clear.

Reply:
ok